api/converters.py

```diff
@@ -201,5 +202,252 @@ def feature_to_legacy_json(f: Feature) -> dict[str, Any]:
+  """Returns a string interpretation of a datetime object, or None."""
+  return str(date) if date is not None else None
+
+def _val_to_list(items: Optional[list]) -> list:
+  """Returns the given list, or returns an empty list if null."""
+  return items if items is not None else []
+
+def _stage_attr(
+    stage: Optional[Stage], field: str, is_mstone: bool=False) -> Optional[Any]:
+  """Returns a specified field of a Stage entity."""
+  if stage is None:
+    return None
+  if is_mstone:
+    return getattr(stage.milestones, field)
+  return getattr(stage, field)
+
+# These FeatureEntry fields should be copied over as basic properties
+# without additional work when converting to JSON.
+BASIC_FE_FIELDS = [
+    "activation_risks",
+    "all_platforms",
+    "all_platforms_descr",
+    "anticipated_spec_changes",
+    "api_spec",
+    "debuggability",
+    "deleted",
+    "ergonomics_risks",
+    "feature_notes",
+    "flag_name",
+    "interop_compat_risks",
+    "measurement",
+    "motivation",
+    "name",
+    "non_oss_deps",
+    "ongoing_constraints",
+    "requires_embedder_support",
+    "security_risks",
+    "star_count",
+    "summary",
+    "tag_review",
+    "unlisted",
+    "webview_risks",
+    "wpt",
+    "wpt_descr"]
+
+def prep_stage_gate_info(
+    fe: FeatureEntry, d: dict) -> dict[str, Optional[Stage]]:
+  """Adds stage and gate info to the dict and returns major stage info."""
+  proto_type = STAGE_TYPES_PROTOTYPE[fe.feature_type]
+  dev_trial_type = STAGE_TYPES_DEV_TRIAL[fe.feature_type]
+  ot_type = STAGE_TYPES_ORIGIN_TRIAL[fe.feature_type]
+  extend_type = STAGE_TYPES_EXTEND_ORIGIN_TRIAL[fe.feature_type]
+  ship_type = STAGE_TYPES_SHIPPING[fe.feature_type]
+
+  stages = Stage.query(Stage.feature_id == d['id'])
+  gates = Gate.query(Gate.feature_id == d['id'])
+  major_stages: dict[str, Optional[Stage]] = {
+      'proto': None,
+      'dev_trial': None,
+      'ot': None,
+      'extend': None}
+
+  # Write a collection of stages and gates associated with the feature,
+  # sorted by type.
+  d['stages'] = {}
+  d['gates'] = collections.defaultdict(list)
+  # Stages and gates are given as a dictionary, with the type as the key,
+  # and a list of entity IDs as the value.
+  for s in stages:
+    if s.stage_type == proto_type:
+      major_stages['proto'] = s
+    elif s.stage_type == dev_trial_type:
+      major_stages['dev_trial'] = s
+    elif s.stage_type == ot_type:
+      major_stages['ot'] = s
+    elif s.stage_type == extend_type:
+      major_stages['extend'] = s
+    elif s.stage_type == ship_type:
+      major_stages['ship'] = s
+    d['stages'][s.stage_type] = s.key.integer_id()
+  for g in gates:
+    d['gates'][g.gate_type].append(g.key.integer_id())
+  
+  return major_stages
+
+def feature_entry_to_json_verbose(fe: FeatureEntry) -> dict[str, Any]:
+  """Returns a verbose dictionary with all info about a feature."""
+  d: dict[str, Any] = {}
+  for field in BASIC_FE_FIELDS:
```

Comment:
Rather than maintaining BASIC_FE_FIELDS as we add fields, could we use ndb's `fe._to_dict()` and then overwrite the items that have a format that we don't like?

Reply:
Agreed. Refactored to do this. 👍